### PR TITLE
Add CUDA/HIP RCM kernels

### PR DESCRIPTION
**Pull request by upsj**
This adds Thrust-based kernels for RCM.

TODO:
- [x] Tests
- [x] Fix OpenMP and Reference handling of connected components (or tests?)

### CODE REVIEW

**Comment by MarcelKoch** [comment]
Could you rebase this? There seem to be a lot of unrelated changes.

**Comment by upsj** [comment]
@MarcelKoch not until #1485 is resolved

File: common/cuda_hip/reorder/rcm_kernels.hpp.inc
Comment (MarcelKoch) [inline common/cuda_hip/reorder/rcm_kernels.hpp.inc:476]: This is not used. So the `construct_inverse_permutation` factory parameter is effectively ignored.

File: hip/components/memory.hip.hpp
Comment (MarcelKoch) [inline hip/components/memory.hip.hpp:150]: This is the same as for `load_relaxed_shared`. I guess HIP doesn't have a finer granularity to distinguish those.

File: test/reorder/rcm.cpp
Comment (MarcelKoch) [inline test/reorder/rcm.cpp:34]: unused
```suggestion
```

File: test/reorder/rcm.cpp
Comment (MarcelKoch) [inline test/reorder/rcm.cpp:84]: `permutation` is unused

File: test/reorder/rcm.cpp
Comment (MarcelKoch) [inline test/reorder/rcm.cpp:128]: conversion warning (`size_type -> int`)

File: test/reorder/rcm.cpp
Comment (MarcelKoch) [inline test/reorder/rcm.cpp:183]: conversion warning (`size_type -> int`)

**Review by MarcelKoch — CHANGES_REQUESTED** [review]
Mostly LGTM. I left some minor comments.

One larger thing is that the cuda/hip implementations ignore the `construct_inverse_permutation` factory parameter. Either that should be deprecated, or added to cuda/hip.

File: common/cuda_hip/reorder/rcm_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/reorder/rcm_kernels.hpp.inc:326]: I think casting a value to a type that cannot represent it is UB, but I'm not 100% on that. Soon, I will add `atomic_cas` functions with signed integers myself that work without this `memcpy` approach with inline assembly.

File: hip/components/memory.hip.hpp
Comment (upsj) [reply hip/components/memory.hip.hpp:150]: they might have a different granularity on the code generation side, since I assume there are dedicated instructions for share memory access, and pointers are likely annotated with `__shared__` internally.

File: common/cuda_hip/reorder/rcm_kernels.hpp.inc
Comment (MarcelKoch) [reply common/cuda_hip/reorder/rcm_kernels.hpp.inc:326]: From cppreference, the conversion signed to unsigned integer is _implementation_ defined (at least until c++20). So it should be valid here. But it's not important.

**Review by MarcelKoch — APPROVED** [review]

File: common/cuda_hip/reorder/rcm_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/reorder/rcm_kernels.hpp.inc:326]: For the CAS to work correctly, I need them to be bitwise identical, until two's complement is standardized, and because this is a temporary solution until we have nice well-typed atomics, I would like to leave the safe version.

File: omp/reorder/rcm_kernels.cpp
Comment (upsj) [inline omp/reorder/rcm_kernels.cpp:372]: this seems to be choosing a max degree node instead of a min degree node? cc @lksriemer

**Review by pratikvn — APPROVED** [review]
Only a few minor comments, otherwise looks good. I think there could be some benefit in using a workspace rather than allocating memory in the kernels, but maybe that can also be done in another PR.

**Review by yhmtsai — COMMENTED** [review]
1/2 part

File: core/test/accessor/scaled_reduced_row_major_reference.cpp
Comment (upsj) [reply core/test/accessor/scaled_reduced_row_major_reference.cpp:56]: it does, but the tests are throwing a lot of warnings for `abs`, and we don't use them for unsigned types, so it's easier to spot warnings if we get rid of these spurious warnings.

File: hip/components/memory.hip.hpp
Comment (upsj) [reply hip/components/memory.hip.hpp:280]: `thrust::complex` is aligned by `2 * sizeof(T)` (differing from `std::complex`, which is only aligned by `sizeof(T)`), and `cudaMalloc` always returns at least 16 byte aligned pointers. The cast is well-defined, see "Array-oriented access" (https://en.cppreference.com/w/cpp/numeric/complex)

File: core/reorder/rcm.cpp
Comment (upsj) [reply core/reorder/rcm.cpp:46]: agreed

File: core/test/accessor/math.cpp
Comment (upsj) [reply core/test/accessor/math.cpp:25]: warnings for `abs(...)` on unsigned types

File: test/reorder/CMakeLists.txt
Comment (upsj) [reply test/reorder/CMakeLists.txt:6]: true, I'll add it

File: test/reorder/rcm.cpp
Comment (upsj) [reply test/reorder/rcm.cpp:105]: yes, for the minimum degree strategy, we only need a minimum degree node from the connected component.

File: omp/reorder/rcm_kernels.cpp
Comment (upsj) [reply omp/reorder/rcm_kernels.cpp:372]: I fixed the issue, but it's pretty hard to test, so I'll just leave it as-is.

**Comment by upsj** [comment]
@yhmtsai I'll mark this as ready to merge already, but wait for the second half of your review before merging

File: common/cuda_hip/reorder/rcm_kernels.hpp.inc
Comment (upsj) [inline common/cuda_hip/reorder/rcm_kernels.hpp.inc:552]: we can do this sorting step once outside of the loop (followed by a level sorting in the end) to get much fewer kernel invocations.

File: common/cuda_hip/reorder/rcm_kernels.hpp.inc
Comment (yhmtsai) [inline common/cuda_hip/reorder/rcm_kernels.hpp.inc:290]: I know bfs should be bradth-first search, but what's the meaning of `u` here?

File: test/reorder/rcm.cpp
Comment (yhmtsai) [inline test/reorder/rcm.cpp:314]: it is not shuffled

**Review by yhmtsai — COMMENTED** [review]
finish my second part. I do not check whether it always gives the same result as the reference. the main question is the shuffled is commented currently

File: test/reorder/rcm.cpp
Comment (upsj) [reply test/reorder/rcm.cpp:314]: oh good catch, I disabled it when debugging and forgot to reenable it

File: common/cuda_hip/reorder/rcm_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/reorder/rcm_kernels.hpp.inc:290]: unordered, since we use atomics to write out the next level and sort afterwards

**Review by yhmtsai — APPROVED** [review]

**Pull request merged**
